src/main/java/sparksoniq/jsoniq/runtime/iterator/primary/ObjectConstructorRuntimeIterator.java

```diff
@@ -75,25 +77,23 @@ public ObjectItem next() {
                 for (RuntimeIterator valueIterator : this._values) {
                     List<Item> currentResults = new ArrayList<>();
                     valueIterator.open(this._currentDynamicContext);
-                    while (valueIterator.hasNext())
+                    while (valueIterator.hasNext()) {
                         currentResults.add(valueIterator.next());
-//                    if(valueIterator.hasNext())
-//                        throw new IteratorFlowException("Object value must return one item!");
+                    }
                     valueIterator.close();
                     //SIMILAR TO ZORBA, if value is more than one item, wrap it in an array
-                    if (currentResults.size() > 1)
+                    if (currentResults.size() > 1) {
                         values.add(new ArrayItem(currentResults));
-                    else
+                    } else if (currentResults.size() == 1) {
                         values.add(currentResults.get(0));
+                    } else {
+                        values.add(new NullItem());
+                    }
                 }
 
                 for (RuntimeIterator keyIterator : this._keys) {
                     keyIterator.open(this._currentDynamicContext);
-                    try {
-                        keys.add(((StringItem) keyIterator.next()).getStringValue());
-                    } catch (Exception ex) {
-                        throw new IteratorFlowException("Object must have string keys!", getMetadata());
-                    }
+                    keys.add((keyIterator.next()).getStringValue());
```

Comment:
Correct use of the item API is that one needs to check that the key is a string by calling isString() here explicitly, and throw the UnexpectedTypeException here in case it is not.

The contract of use of the item API, is that getStringValue() may only be called after calling isString() and its returning true. The runtime exceptions in the item API are internal to the development (they are not supposed to happen, ever -- if they ever do it means that there is a bug in our code) and are not meant be (ab)used to throw external type errors.

This means that the renaming of RuntimeException to UnexpectedTypeException should be reverted.

Thanks a lot!

Reply:
That makes a lot of sense. I will be updating the code accordingly. Thanks for the explanation!